### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by marouaneamz**

### CODE REVIEW

File: tests/test_evaluation/test_metrics/test_multi_metrics.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_metrics.py]: please rename this, `test_multi_task_metrics.py`

File: tests/test_evaluation/test_metrics/test_multi_metrics.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import copy
+from unittest import TestCase
+
+import numpy as np
+import torch
+
+from mmcls.evaluation.metrics import Accuracy
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_metrics.py]: The metric you are testing should not be 'Accuracy' it should be 
'MultiTaskMetric', no ?

```suggestion
from mmcls.evaluation.metrics import MultiTaskMetric
```

File: tests/test_evaluation/test_metrics/test_multi_metrics.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    },
+                ], [{'task0': 0, 'task1': 2}, {'task0': 2, 'task1': 1}])
+        ]
+
+        # Test with score (use score instead of label if score exists)
+        metric = METRICS.build(dict(type='Accuracy', thrs=0.6))
+        metric.process(None, pred)
+        acc = metric.evaluate(2)
+        self.assertIsInstance(acc, dict)
+        self.assertAlmostEqual(acc['accuracy/top1'], 2 / 6 * 100, places=4)
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_metrics.py]:
```suggestion
        self.assertAlmostEqual(acc['task0_accuracy/top1'], 2 / 6 * 100, places=4)
```

**Review by piercus — CHANGES_REQUESTED** [review]

File: tests/test_evaluation/test_metrics/test_multi_task_metrics.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    task_metrics = {
+            'task0': [dict(type='Accuracy', topk=(1, ))],
+            'task1': [
+                dict(type='Accuracy', topk=(1, 3)),
+                dict(type='SingleLabelMetric', items=['precision', 'recall'])
+                ]
+            }
+
+    def test_init(self):
+        metrics = MultiTasksMetric(self.task_metrics)
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_task_metrics.py]:
```suggestion
        metric = MultiTasksMetric(self.task_metrics)
```

File: tests/test_evaluation/test_metrics/test_multi_task_metrics.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        self.assertIsInstance(metrics.Accuracy_task0, Accuracy)
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_task_metrics.py]: * I won't put it in the root metric object
* Better to name it task0_<something>
* Accuracy is strange to have in the key here
* it should be a list of metrics for each task

File: tests/test_evaluation/test_metrics/test_multi_task_metrics.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                'gt_task': {'task0': 0, 'task1': 2},
+                'pred_task': {
+                    'task0': torch.tensor([0.0, 0.0, 1.0]),
+                    'task1': torch.tensor([0.0, 0.0, 1.0])
+                },
+             }
+        ]
+        output = MultiTasksMetric(
+            task_metrics=self.task_metrics).compute_metrics(results)
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_task_metrics.py]: No need to test compute_metric here : 
* compute_metric is an internal method
* those tests arel already done in test_evaluate

**Review by piercus — CHANGES_REQUESTED** [review]

File: tests/test_evaluation/test_metrics/test_multi_task_metrics.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        self.assertIsInstance(output, dict)
+        self.assertAlmostEqual(output['task0_accuracy/top1'], 1 / 2 * 100)
+        self.assertGreater(output['task1_precision/top1'], 0)
+
+    def test_evaluate(self):
+        """Test using the metric in the same way as Evalutor."""
+
+        # Test with score (use score instead of label if score exists)
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_task_metrics.py:91]: Can you please test more cases : 
* nested tasks
* multiple metric for each task
* task without any ground truth value 

...

**Review by piercus — CHANGES_REQUESTED** [review]

File: tests/test_evaluation/test_metrics/test_multi_task_metrics.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        self.assertIsInstance(results, dict)
+        self.assertAlmostEqual(results['task0_accuracy/top1'],  100)
+        self.assertGreater(results['task1_single-label/precision'], 0)
+
+        # Test nested
+        metric = MultiTasksMetric(self.task_metrics2)
+        metric.process(None, self.pred2)
+        results = metric.evaluate(2)
+        self.assertIsInstance(results, dict)
+        self.assertGreater(results['task10_single-label/precision'], 0)
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_task_metrics.py]:
```suggestion
        self.assertGreater(results['task1_task10_single-label/precision'], 0)
```

File: tests/test_evaluation/test_metrics/test_multi_task_metrics.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        self.assertGreater(results['task11_accuracy/top1'], 0)
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_task_metrics.py]:
```suggestion
        self.assertGreater(results['task1_task11_accuracy/top1'], 0)
```

**Review by piercus — CHANGES_REQUESTED** [review]

File: tests/test_evaluation/test_metrics/test_multi_task_metrics.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        }, {
+            'task0': 2,
+            'task1': {
+                'task10': 1,
+                'task11': 0
+            }
+        }])
+    ]
+
+    pred3 = [MultiTaskDataSample()]
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_task_metrics.py]:
```suggestion
    pred3 = [MultiTaskDataSample().to_dict()]
```

File: mmcls/evaluation/metrics/multi_task.py [hunk 11]
```diff
@@ -2,24 +2,85 @@
...
 
+        self._metrics = {}
         for task_name in self.task_metrics.keys():
-            globals()['metric_%s' % task_name] = mmengineMetrics.build(
-                self.task_metrics[task_name])
+            if type(self.task_metrics[task_name]) == list:
+                self._metrics[task_name] = []
+                for metric in self.task_metrics[task_name]:
+                    self._metrics[task_name].append(METRICS.build(metric))
+            elif type(self.task_metrics[task_name]) == dict:
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: you need to implement a recursive strategy here, it will fail if you have a 3-level nested tasks

File: mmcls/evaluation/metrics/multi_task.py [hunk 12]
```diff
@@ -30,53 +91,62 @@ def process(self, data_batch, data_samples: Sequence[dict]):
...
-                        raise Exception('score masked')
-                    data_sample['pred_label']['score'] = data_sample[
-                        'pred_label']['score'][index]
-                    task_data_samples.append(data_sample)
-            globals()['metric_%s' % task_name].process(data_batch,
-                                                       task_data_samples)
+            if type(self.task_metrics[task_name]) != dict:
+                task_data_sample_dicts = self.pre_process_cls(
+                    data_sample_instances, task_name)
+                # à reflichir
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: french ?

File: mmcls/evaluation/metrics/multi_task.py [hunk 13]
```diff
@@ -30,53 +91,62 @@ def process(self, data_batch, data_samples: Sequence[dict]):
...
+                for metric in self._metrics[task_name]:
+                    metric.process(data_batch, task_data_sample_dicts)
+            else:
+                task_data_sample = self.pre_process_nested(
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: recursivity is needed here

**Review by piercus — CHANGES_REQUESTED** [review]

File: mmcls/evaluation/metrics/multi_task.py [hunk 14]
```diff
@@ -2,24 +2,78 @@
...
+        task_metrics2 = {
+            'task0': [dict(type='Accuracy', topk=(1, ))],
+            'task1': {
+                'task10': [
+                    dict(type='Accuracy', topk=(1, 3)),
+                    dict(type='SingleLabelMetric', items=['precision'])
+                ],
+                'task11': [dict(type='Accuracy', topk=(1, ))]
+            }
+        }
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: Please rework the example as in https://github.com/marouaneamz/mmclassification/blob/5d26026f4b1f410df847c787fcabf90cee63473e/mmcls/evaluation/metrics/single_label.py#L75-L101

File: mmcls/evaluation/metrics/multi_task.py [hunk 15]
```diff
@@ -30,53 +84,64 @@ def process(self, data_batch, data_samples: Sequence[dict]):
...
-            globals()['metric_%s' % task_name].process(data_batch,
-                                                       task_data_samples)
+            filtered_data_samples = []
+            for data_sample in data_sample_instances:
+                sample_mask = data_sample.get_task_mask(task_name)
+                if sample_mask:
+                    filtered_data_samples.append(data_sample)
+            if type(self.task_metrics[task_name]) != dict:
+                for metric in self._metrics[task_name]:
+                    if metric.__class__.__name__ != 'MultiTasksMetric':
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py:125]: I think we should add a comment here
```suggestion
                    # Current implementation is only comptaible 
                    # With 2 types of metrics : 
                    # * Cls Metrics 
                    # * Nested Cls Metrics
                    # In order to make it work with other
                    # non-cls heads/metrics, one will have to 
                    # override the current implementation
                    if metric.__class__.__name__ != 'MultiTasksMetric':
```

File: mmcls/evaluation/metrics/multi_task.py [hunk 16]
```diff
@@ -30,53 +84,64 @@ def process(self, data_batch, data_samples: Sequence[dict]):
...
+            size (int): Length of the entire validation dataset. When batch
+                size > 1, the dataloader may pad some data samples to make
+                sure all ranks have the same length of dataset slice. The
+                ``collect_results`` function will drop the padded data based on
+                this size.
         Returns:
-            Dict: The computed metrics. The keys are the names of the metrics,
-            and the values are corresponding results.
+            dict: Evaluation metrics dict on the val dataset. The keys are
+            task_name + the names of the metrics, and the values
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]:
```suggestion
            "{task_name}_{metric_name}", and the values
```

**Review by piercus — CHANGES_REQUESTED** [review]
documentation related feedback

File: mmcls/evaluation/metrics/multi_task.py [hunk 17]
```diff
@@ -2,24 +2,107 @@
...
+        task_metrics2 = {
+            'task0': [dict(type='Accuracy', topk=(1, ))],
+            'task1': {
+                'task10': [
+                    dict(type='Accuracy', topk=(1, 3)),
+                    dict(type='SingleLabelMetric', items=['precision'])
+                ],
+                'task11': [dict(type='Accuracy', topk=(1, ))]
+            }
+        }
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: useless

File: mmcls/evaluation/metrics/multi_task.py [hunk 18]
```diff
@@ -2,24 +2,107 @@
...
+    Examples:
+        >>> import torch
+        >>> from mmcls.evaluation import MultiTasksMetric
+        >>> # -------------------- The Basic Usage --------------------
+        >>>task_metrics = {
+            'task0': [dict(type='Accuracy', topk=(1, ))],
+            'task1': [
+                dict(type='Accuracy', topk=(1, 3)),
+                dict(type='SingleLabelMetric', items=['precision', 'recall'])
+            ]
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: Make it simple like

```suggestion
            'task1': [dict(type='Accuracy', topk=(1, 3))]
```

File: mmcls/evaluation/metrics/multi_task.py [hunk 19]
```diff
@@ -2,24 +2,107 @@
...
+                    'task1': torch.tensor([0.0, 0.0, 1.0])
+                },
+            ], [{
+                'task0': 0,
+                'task1': 2
+            }, {
+                'task0': 2,
+                'task1': 2
+            }])
+        ]
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: to simplify

File: mmcls/evaluation/metrics/multi_task.py [hunk 20]
```diff
@@ -2,24 +2,107 @@
...
+        >>>metric = MultiTasksMetric(self.task_metrics)
+        >>>metric.process(None, self.pred)
+        >>>metric.evaluate(2)
+        {'task0_accuracy/top1': 100.0, 'task1_accuracy/top1': 50.0,
+        'task1_accuracy/top3': 100.0, 'task1_single-label/precision':
+         33.33333206176758, 'task1_single-label/recall': 16.66666603088379}
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: more clear indentation

**Review by piercus — CHANGES_REQUESTED** [review]

File: mmcls/evaluation/metrics/multi_task.py [hunk 21]
```diff
@@ -2,24 +2,90 @@
...
+        >>> import torch
+        >>> from mmcls.evaluation import MultiTasksMetric
+        >>> # -------------------- The Basic Usage --------------------
+        >>>task_metrics = {
+            'task0': [dict(type='Accuracy', topk=(1, ))],
+            'task1': [dict(type='Accuracy', topk=(1, 3)),]
+        }
+        >>>pred = [
+            {
+            'gt_tassk'{
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]:
```suggestion
            'gt_task'{
```

File: mmcls/evaluation/metrics/multi_task.py [hunk 22]
```diff
@@ -2,24 +2,90 @@
...
+                'task0': torch.tensor([0.7, 0.0, 0.3]),
+                'task1': torch.tensor([0.5, 0.2, 0.3])
+                }
+            'pred_task' : {'task0:0 , task2:2'}
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: It's strange to have gt : tensor and pred : not tensor, is it the opposite ?

File: mmcls/evaluation/metrics/multi_task.py [hunk 23]
```diff
@@ -2,24 +2,90 @@
...
+                }
+            'pred_task' : {'task0:2 , task2:2'}
+            },
+        ]
+        >>>metric = MultiTasksMetric(self.task_metrics)
+        >>>metric.process(None, self.pred)
+        >>>metric.evaluate(2)
+        {'task0_accuracy/top1': 100.0,
+        'task1_accuracy/top1': 50.0,
+        'task1_accuracy/top3': 100.0}
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]:
```suggestion
        {
          'task0_accuracy/top1': 100.0,
          'task1_accuracy/top1': 50.0,
          'task1_accuracy/top3': 100.0
        }
```

File: mmcls/evaluation/metrics/multi_task.py [hunk 24]
```diff
@@ -2,24 +2,90 @@
...
+        }
+        >>>pred = [
+            {
+            'gt_tassk'{
+                'task0': torch.tensor([0.7, 0.0, 0.3]),
+                'task1': torch.tensor([0.5, 0.2, 0.3])
+                }
+            'pred_task' : {'task0:0 , task2:2'}
+            },
+            {'gt_task':
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: Please clean up
```suggestion
            {
              'gt_task':
```

File: mmcls/evaluation/metrics/multi_task.py [hunk 25]
```diff
@@ -2,24 +2,90 @@
...
+                'task0': torch.tensor([0.0, 0.0, 1.0]),
+                'task1': torch.tensor([0.0, 0.0, 1.0])
+                }
+            'pred_task' : {'task0:2 , task2:2'}
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: clean up
```suggestion
                },
                'pred_task' : {'task0:2 , task2:2'}
```

File: mmcls/evaluation/metrics/multi_task.py [hunk 26]
```diff
@@ -2,24 +2,90 @@
...
+            },
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py:37]:
```suggestion
            }
```

File: mmcls/evaluation/metrics/multi_task.py [hunk 27]
```diff
@@ -2,24 +2,90 @@
...
+        >>>task_metrics = {
+            'task0': [dict(type='Accuracy', topk=(1, ))],
+            'task1': [dict(type='Accuracy', topk=(1, 3)),]
+        }
+        >>>pred = [
+            {
+            'gt_tassk'{
+                'task0': torch.tensor([0.7, 0.0, 0.3]),
+                'task1': torch.tensor([0.5, 0.2, 0.3])
+                }
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]:
```suggestion
                },
```

File: mmcls/evaluation/metrics/multi_task.py (hunk 25 above)
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py]: It's strange to have gt : tensor and pred : not tensor, is it the opposite ?

**Review by piercus — CHANGES_REQUESTED** [review]

**Review by piercus — APPROVED** [review]
